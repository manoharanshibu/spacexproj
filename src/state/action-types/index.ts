export enum ActionType {
    GET_ITEMS_SUCCESS = "getitemssuccess",
}