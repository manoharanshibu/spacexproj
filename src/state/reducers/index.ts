import { combineReducers } from 'redux';
import itemsReducer from './itemsReducer';

const reducers = combineReducers({
    itemsReducer: itemsReducer
})

export default reducers;

export type State = ReturnType<typeof reducers>