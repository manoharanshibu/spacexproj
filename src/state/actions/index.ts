import { ActionType } from '../action-types/index';

interface IGetItemsSuccessAction {
    type: ActionType.GET_ITEMS_SUCCESS
    payload: any
}

export type Action = IGetItemsSuccessAction