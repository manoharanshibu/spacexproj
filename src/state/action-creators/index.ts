import { ActionType } from "../action-types"
import { Action } from "../actions"
import axios from 'axios';
import { Dispatch } from 'redux';

const endpoint = process.env.REACT_APP_SPACEX_END_POINT as string;
const config = {
    headers: {
    },
};

export const fetchItems = () => {
    return async (dispatch: Dispatch<Action>) => {
        const response = await axios.get(`${endpoint}/`, config);
        try {
            dispatch({
                type: ActionType.GET_ITEMS_SUCCESS,
                payload: response ? response.data : []
            });
        } catch(error) {
            console.error(error);
        }
    }
}
