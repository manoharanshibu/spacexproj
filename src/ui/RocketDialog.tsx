import * as React from 'react';
import { useState, useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { State } from '../state';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogContentText from '@material-ui/core/DialogContentText';
import DialogTitle from '@material-ui/core/DialogTitle';
import Draggable from 'react-draggable';
import { ActionType } from '../state/action-types';
import Paper from '@material-ui/core/Paper';

import Table from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableCell from '@mui/material/TableCell';
import TableContainer from '@mui/material/TableContainer';
import TableRow from '@mui/material/TableRow';
import './RocketDialog.scss';

const PaperComponent = (props: any) => {
    return (
        <Draggable
            handle="#draggable-dialog-title"
            cancel={'[class*="MuiDialogContent-root"]'}
        >
            <Paper {...props} />
        </Draggable>
    );
}

interface IProps {
    open: boolean
}

const RocketDialog: React.FC<IProps> = (props) => {
    const [open, setOpen] = useState(props.open);
    const dispatch = useDispatch();
    const { showRocketDetails, selectedRocket } = useSelector((state: State) => state.itemsReducer);

    useEffect(() => {
        setOpen(showRocketDetails);
    }, [showRocketDetails])

    const handleClose = () => {
        setOpen(false);
        dispatch({
            type: ActionType.SHOW_ROCKET_DETAILS,
            payload: {
                showRocketDetails: false,
                selectedRocket: null
            }
        });
    };

    const getDetails = () => {
        if (selectedRocket) {
            return <TableContainer component={Paper}>
                <Table sx={{ minWidth: 250 }} aria-label="simple table">
                    <TableBody>
                        <TableRow
                            sx={{ '&:last-child td, &:last-child th': { border: 0 } }}
                        >
                            <TableCell>Rocket ID</TableCell>
                            <TableCell>{selectedRocket.rocket_id}</TableCell>
                        </TableRow>
                        <TableRow
                            sx={{ '&:last-child td, &:last-child th': { border: 0 } }}
                        >
                            <TableCell>Name</TableCell>
                            <TableCell>{selectedRocket.rocket_name}</TableCell>
                        </TableRow>
                        <TableRow
                            sx={{ '&:last-child td, &:last-child th': { border: 0 } }}
                        >
                            <TableCell>Type</TableCell>
                            <TableCell>{selectedRocket.rocket_type}</TableCell>
                        </TableRow>
                    </TableBody>
                </Table>
            </TableContainer>
        } else {
            return <div>No Data</div>
        }
    }

    return (
        <Dialog
            open={open === true ? true : false}
            onClose={handleClose}
            PaperComponent={PaperComponent}
            aria-labelledby="draggable-dialog-title"
        >
            <DialogTitle className='dialog-title' id="draggable-dialog-title" data-testid="draggable-dialog-title">
                Rocket Details
                </DialogTitle>
            <DialogContent>
                <DialogContentText>
                    {getDetails()}
                </DialogContentText>
            </DialogContent>
            <DialogActions>
                <Button autoFocus onClick={handleClose}>
                    Close
                    </Button>
            </DialogActions>
        </Dialog>
    );
}

export default RocketDialog;
