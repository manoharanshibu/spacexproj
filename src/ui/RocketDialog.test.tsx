import React from 'react';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux'
import { store } from './../state/index';
import RocketDialog from '../ui/RocketDialog';
import renderer from 'react-test-renderer';

const initialState = {
    items: [],
    loadedItems: false,
    showRocketDetails: true,
    selectedRocket: null
}

jest.mock('react-redux', () => ({
    useDispatch: () => { },
    useSelector: () => ({
        state: initialState,
    }),
}));

describe(('Rocket Dialog Component'), () => {
    it('renders correctly', () => {
        const tree = renderer
            .create(<RocketDialog open={true} />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
})
