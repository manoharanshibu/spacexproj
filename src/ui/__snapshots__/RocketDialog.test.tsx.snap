// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rocket Dialog Component renders correctly 1`] = `null`;
